### PR TITLE
[SMTChecker] Support array length

### PR DESCRIPTION
**Pull request by leonardoalt**
Refs https://github.com/ethereum/solidity/issues/6048

Missing:
- [x] CVC4 `tuple_constructor`
- [x] smtlib2 `tuple_constructor`
- [x] Rerun tests then

### CODE REVIEW

File: libsolidity/formal/SMTEncoder.cpp
Comment (leonardoalt) [inline libsolidity/formal/SMTEncoder.cpp:1663]: @ekpyron this was the problem on the BMC side (which uses the `mergeVariables` function).
The index of the symbolic array was being reset after a branch, but not of the tuple.

Comment (leonardoalt) [reply libsolidity/formal/SMTEncoder.cpp:1663]: The idea here is that we need to set the current index but keep the `next` as it is, without zeroing, because we still need to `ite` (merge) after the branches. That's why I added the function `SSAVariable::setIndex` that does that. The old `SSAVariable::reset` zeroes the current index and sets next to 1, effectively resetting.

**Review by ekpyron — CHANGES_REQUESTED** [review]
Tests need fixing. Where did those failures come from? Will still have a look at the code as well and leave another review in a bit.

File: libsolidity/formal/SMTEncoder.cpp
Comment (ekpyron) [inline libsolidity/formal/SMTEncoder.cpp:911]: Can't you trigger this with the following?
```
library L { function f(uint256[] memory) public pure {} }
contract C { using L for uint256[]; function g() public pure { uint256[] memory x; x.f(); } }
```
Or something like that?

**Comment by leonardoalt** [comment]
Apparently what happens in those tests is that z3 times out and CVC4 solves it. I think they should be updated to my latest run with both solvers... maybe I should set them as cvc4 only

**Comment by leonardoalt** [comment]
Because the CI version might not be running z3, so it doesn't get the error message for it timing out

File: libsolidity/formal/SMTEncoder.cpp
Comment (ekpyron) [inline libsolidity/formal/SMTEncoder.cpp:969]: Hm... unrelated to this PR, but I wonder whether it makes sense to have a look at explicit ownership vs all the shared pointers in the smt code at some point...

File: libsolidity/formal/SMTEncoder.cpp
Comment (ekpyron) [inline libsolidity/formal/SMTEncoder.cpp:1371]: So what happens in the ``else`` case?

File: libsolidity/formal/SMTEncoder.cpp
Comment (ekpyron) [reply libsolidity/formal/SMTEncoder.cpp:1663]: Hm - I'm not sure I understand this entirely, but I'd go on a leap and trust you that this works on this one :-).

**Comment by leonardoalt** [comment]
Ah, those set ups actually don't have cvc4, what?

File: libsolidity/formal/SSAVariable.cpp
Comment (ekpyron) [inline libsolidity/formal/SSAVariable.cpp]: Without thinking about this too much, but from looking at the code in the diff: Why exactly is ``m_nextFreeIndex`` a unique pointer :-)? Unrelated to this PR, though.
But I mean: either this can be unset, then we should at least check or assert here - or it *can't* be unset, then it doesn't need to be a unique pointer, does it?

File: libsolidity/formal/SymbolicVariables.cpp
Comment (ekpyron) [inline libsolidity/formal/SymbolicVariables.cpp:212]: This diff is just mean :-D.

File: libsolidity/formal/SymbolicVariables.h
Comment (ekpyron) [inline libsolidity/formal/SymbolicVariables.h]: I would have gone for calling this ``elements()`` or something like that...

**Review by ekpyron — COMMENTED** [review]
Haven't fully reviewed ``SymbolicVariables.{cpp,h}`` and didn't look at the tests yet, but some comments.

File: libsolidity/formal/SMTEncoder.cpp
Comment (leonardoalt) [reply libsolidity/formal/SMTEncoder.cpp:969]: It does: https://github.com/ethereum/solidity/issues/7266

File: libsolidity/formal/SMTEncoder.cpp
Comment (leonardoalt) [reply libsolidity/formal/SMTEncoder.cpp:1371]: Then the constraint that says that the array/string/bytes has a certain value isn't added, and the array is just a fresh variable.
Maybe I should add a TODO for the else case, where the solution would be to `smt::store` each string literal item in the array

File: libsolidity/formal/SSAVariable.cpp
Comment (leonardoalt) [reply libsolidity/formal/SSAVariable.cpp]: This is for the case where SSAVariables are copied, which happens when there are branches. In that case, we want all the SSAVariables that refer to the same program variable to have the same `next` which must be updated if any of those update it.

File: libsolidity/formal/SymbolicVariables.cpp
Comment (leonardoalt) [reply libsolidity/formal/SymbolicVariables.cpp:212]: Yea, I always select `hide whitespace changes` :p

File: libsolidity/formal/SymbolicVariables.h
Comment (leonardoalt) [reply libsolidity/formal/SymbolicVariables.h]: Calling `array` `elements`? Not sure I get it

**Comment by leonardoalt** [comment]
> Hm - I'm not sure I understand this entirely, but I'd go on a leap and trust you that this works on this one :-).

Re this (GH didn't let me respond there):
If we only assign `index() = foo`, it only assigns the SSA index of the `SymbolicVariable`. However, `SymbolicFunctionVariable` and `SymbolicArrayVariable` have extra `SymbolicVariable`s as members, so they need to relay that assignment to their members as well. That's why I made it a function.

File: libsolidity/formal/SymbolicVariables.h
Comment (ekpyron) [reply libsolidity/formal/SymbolicVariables.h]: Just a different name for the function... I mean this thing is a symbolic array - it's a bit weird that it has a member called ``array``... so personally I'd call the function ``elements()`` - i.e. you query the expression corresponding to the array elements, not to the array length...

Then you'd not have ``select(array->array(), index)``, but ``select(array->elements(), index)``, etc...
But that's just a side-node about personal taste - I'm fine with keeping this be called ``array`` :-).

File: libsolidity/formal/SSAVariable.cpp
Comment (ekpyron) [reply libsolidity/formal/SSAVariable.cpp]: But - by definition - you can't share a unique pointer!

Comment (ekpyron) [reply libsolidity/formal/SSAVariable.cpp]: Can you point me to the code that does this? If you have two unique_ptr's containing the same pointer, then you'll end up with double-frees...

Comment (ekpyron) [reply libsolidity/formal/SSAVariable.cpp]: I mean - an SSAVariable with a unique_ptr member won't be copiable, so apparently SSAVariables *aren't* copied... and the ``m_nextFreeIndex`` is private and only used by dereferencing or by assigning to ``= make_unique<unsigned>(1);``... to me it very much looks like this could just be a plain member ``unsigned m_nextFreeIndex = 1;`` :-).

File: libsolidity/formal/SymbolicVariables.cpp
Comment (ekpyron) [reply libsolidity/formal/SymbolicVariables.cpp:212]: Doesn't help for this one, though :-).

File: libsolidity/formal/SymbolicVariables.h
Comment (leonardoalt) [reply libsolidity/formal/SymbolicVariables.h]: Ah, right. I thought you meant the tuple elements. Fine with changing.

File: libsolidity/formal/SSAVariable.cpp
Comment (leonardoalt) [reply libsolidity/formal/SSAVariable.cpp]: Damn, true. Then I'd rather make it a `shared_ptr` though

Comment (leonardoalt) [reply libsolidity/formal/SSAVariable.cpp]: Tbh then I don't know how it's working right now haha

Comment (ekpyron) [reply libsolidity/formal/SSAVariable.cpp]: Yeah - either this doesn't need to be a pointer at all, or there's something fishy about this, if you think it has to be :-D.

File: libsolidity/formal/SMTEncoder.cpp
Comment (leonardoalt) [reply libsolidity/formal/SMTEncoder.cpp:911]: It wasn't triggered, seems fine, but I added an assertion that didn't work...

File: libsolidity/formal/SSAVariable.cpp
Comment (leonardoalt) [reply libsolidity/formal/SSAVariable.cpp]: Removed pointer, let's see

**Comment by leonardoalt** [comment]
@ekpyron updated with suggestions.
Tests still failing because of the static/dynamic z3 stuff

**Comment by leonardoalt** [comment]
Not sure why tests are still failing

**Comment by leonardoalt** [comment]
@ekpyron ready for review again

**Comment by leonardoalt** [comment]
Rebased

File: libsolidity/formal/SMTEncoder.cpp
Comment (ekpyron) [reply libsolidity/formal/SMTEncoder.cpp:911]: Ah, I now see it - the ``if (accessType->category() == Type::Category::Function)`` above catches these I think, so it's fine!

Comment (ekpyron) [reply libsolidity/formal/SMTEncoder.cpp:911]: Although, it might be worth to check/verify that this either actually works or produces a warning, but that we can do later.

Comment (leonardoalt) [reply libsolidity/formal/SMTEncoder.cpp:911]: I'd actually suggest removing the `else` completely, since https://github.com/ethereum/solidity/pull/8916 removes it anyway

File: libsolidity/formal/SSAVariable.h
Comment (ekpyron) [inline libsolidity/formal/SSAVariable.h:48]: This is still a bit confusing to me :-). Usually it is just ``m_nextFreeIndex == (m_currentIndex + 1)``... but using the ``unsigned &index()`` you might assign a different value to the current index... but ``nextFreeIndex`` will remain the same... you're saying this works for combining SSA variables somehow? I'll need to look at how that's done at some point :-).
But no need for this PR.

**Review by ekpyron — APPROVED** [review]
I think this is fine now!

**Comment by leonardoalt** [comment]
@ekpyron wait, the `push`/`pop` warning is still there

**Comment by ekpyron** [comment]
> @ekpyron wait, the `push`/`pop` warning is still there

You mean the assertion of the member name? Yes, I figured out why it's not triggered and I think it's fine, so we can keep it. I commented about it and resolved the conversation about it. Or do you mean something else?

EDIT: ah, now I saw your comment about that. If you still want to change that, go for it, but it's also fine as it is.

File: libsolidity/formal/SMTEncoder.cpp
Comment (ekpyron) [reply libsolidity/formal/SMTEncoder.cpp:911]: Well keeping it is fine as well and having the unsupported warning until #8916 is merged doesn't hurt, does it? But I'll also reapprove, if you remove it.

**Comment by leonardoalt** [comment]
Ok, if it's fine as it is I'll avoid rerunning all the tests.

File: libsolidity/formal/SSAVariable.h
Comment (leonardoalt) [reply libsolidity/formal/SSAVariable.h:48]: The constructor calls `resetIndex` which does that. Probably not very clear...  but should work

**Pull request merged**
